luxonis_train/attached_modules/losses/reconstruction_segmentation_loss.py

```diff
@@ -0,0 +1,297 @@
+        (_, channel, _, _) = img1.size()
+
+        if channel == self.channel and self.window.dtype == img1.dtype:
+            window = self.window
+        else:
+            window = (
+                create_window(self.window_size, channel)
+                .to(img1.device)
+                .type(img1.dtype)
+            )
+            self.window = window
+            self.channel = channel
+
+        s_score, ssim_map = ssim(
+            img1,
+            img2,
+            window=window,
+            window_size=self.window_size,
+            size_average=self.size_average,
+        )
+        return 1.0 - s_score
+
+
+def create_window(window_size: int, channel: int = 1) -> Tensor:
+    _1D_window = gaussian(window_size, 1.5).unsqueeze(1)
+    _2D_window = (
+        _1D_window.mm(_1D_window.t()).float().unsqueeze(0).unsqueeze(0)
+    )
+    window = _2D_window.expand(
+        channel, 1, window_size, window_size
+    ).contiguous()
+    return window
+
+
+def gaussian(window_size: int, sigma: float) -> Tensor:
+    gauss = torch.Tensor(
+        [
+            exp(-((x - window_size // 2) ** 2) / float(2 * sigma**2))
+            for x in range(window_size)
+        ]
+    )
+    return gauss / gauss.sum()
+
+
+def ssim(
+    img1: Tensor,
+    img2: Tensor,
+    window_size: int = 11,
+    window: Tensor | None = None,
+    size_average=True,
+    full=False,
+    val_range=None,
+) -> Tensor:
+    if val_range is None:
+        if torch.max(img1) > 128:
+            max_val = 255
+        else:
+            max_val = 1
+
+        if torch.min(img1) < -0.5:
+            min_val = -1
+        else:
+            min_val = 0
+        dynamic_range = max_val - min_val
+    else:
+        dynamic_range = val_range
+
+    padd = window_size // 2
+    (_, channel, height, width) = img1.size()
+    if window is None:
+        real_size = min(window_size, height, width)
+        window = create_window(real_size, channel=channel).to(img1.device)
+
+    mu1 = F.conv2d(img1, window, padding=padd, groups=channel)
+    mu2 = F.conv2d(img2, window, padding=padd, groups=channel)
+
+    mu1_sq = mu1.pow(2)
+    mu2_sq = mu2.pow(2)
+    mu1_mu2 = mu1 * mu2
+
+    sigma1_sq = (
+        F.conv2d(img1 * img1, window, padding=padd, groups=channel) - mu1_sq
+    )
+    sigma2_sq = (
+        F.conv2d(img2 * img2, window, padding=padd, groups=channel) - mu2_sq
+    )
+    sigma12 = (
+        F.conv2d(img1 * img2, window, padding=padd, groups=channel) - mu1_mu2
+    )
+
+    c1 = (0.01 * dynamic_range) ** 2
+    c2 = (0.03 * dynamic_range) ** 2
+
+    v1 = 2.0 * sigma12 + c2
+    v2 = sigma1_sq + sigma2_sq + c2
+    cs = torch.mean(v1 / v2)  # contrast sensitivity
+
+    ssim_map = ((2 * mu1_mu2 + c1) * v1) / ((mu1_sq + mu2_sq + c1) * v2)
+
+    if size_average:
+        ret = ssim_map.mean()
+    else:
+        ret = ssim_map.mean(1).mean(1).mean(1)
+
+    if full:
+        return ret, cs
+    return ret, ssim_map
+
+
+class FocalLoss(nn.Module):
```

Comment:
Can this be replaced by the focal loss already in luxonis-train? Or can we extend the one already implemented to support functionality introduced in this one as well (e.g. label smoothing)?